modules/cudastereo/src/stereosgm.cpp

```diff
@@ -0,0 +1,153 @@
+        StereoSGMImpl(int minDisparity, int numDisparities, int P1, int P2, int uniquenessRatio, int mode);
+
+        void compute(InputArray left, InputArray right, OutputArray disparity) CV_OVERRIDE;
+        void compute(InputArray left, InputArray right, OutputArray disparity, Stream& stream) CV_OVERRIDE;
+
+        int getBlockSize() const CV_OVERRIDE { return 1; }
+        void setBlockSize(int /*blockSize*/) CV_OVERRIDE {}
+
+        int getDisp12MaxDiff() const CV_OVERRIDE { return 1; }
+        void setDisp12MaxDiff(int /*disp12MaxDiff*/) CV_OVERRIDE {}
+
+        int getMinDisparity() const CV_OVERRIDE { return 1; }
+        void setMinDisparity(int /*minDisparity*/) CV_OVERRIDE {}
+
+        int getNumDisparities() const CV_OVERRIDE { return params.numDisparities; }
+        void setNumDisparities(int numDisparities) CV_OVERRIDE { params.numDisparities = numDisparities; }
+
+        int getSpeckleWindowSize() const CV_OVERRIDE { return 0; }
+        void setSpeckleWindowSize(int /*speckleWindowSize*/) CV_OVERRIDE {}
+
+        int getSpeckleRange() const CV_OVERRIDE { return 0; }
+        void setSpeckleRange(int /*speckleRange*/) CV_OVERRIDE {}
+
+        int getP1() const CV_OVERRIDE { return params.P1; }
+        void setP1(int P1) CV_OVERRIDE { params.P1 = P1; }
+
+        int getP2() const CV_OVERRIDE { return params.P2; }
+        void setP2(int P2) CV_OVERRIDE { params.P2 = P2; }
+
+        int getUniquenessRatio() const CV_OVERRIDE { return params.uniquenessRatio; }
+        void setUniquenessRatio(int uniquenessRatio) CV_OVERRIDE { params.uniquenessRatio = uniquenessRatio; }
+
+        int getMode() const CV_OVERRIDE { return MODE_HH; }
+        void setMode(int /*mode*/) CV_OVERRIDE {}
+
+        int getPreFilterCap() const CV_OVERRIDE { return -1; }
+        void setPreFilterCap(int /*preFilterCap*/) CV_OVERRIDE {}
+
+    private:
+        StereoSGMParams params;
+        GpuMat censused_left, censused_right;
+        GpuMat aggregated;
+        GpuMat left_disp_tmp, right_disp_tmp;
+        GpuMat right_disp;
+    };
+
+    StereoSGMImpl::StereoSGMImpl(int minDisparity, int numDisparities, int P1, int P2, int uniquenessRatio, int mode)
+        : params(minDisparity, numDisparities, P1, P2, uniquenessRatio, mode)
+    {
+    }
+
+    void StereoSGMImpl::compute(InputArray left, InputArray right, OutputArray disparity)
+    {
+        compute(left, right, disparity, Stream::Null());
+    }
+
+    void StereoSGMImpl::compute(InputArray _left, InputArray _right, OutputArray _disparity, Stream& _stream)
+    {
+        using namespace ::cv::cuda::device::stereosgm;
+
+        GpuMat left = _left.getGpuMat();
+        GpuMat right = _right.getGpuMat();
+        const Size size = left.size();
+
+        CV_Assert(left.type() == CV_8UC1 || left.type() == CV_16UC1);
+        CV_Assert(size == right.size() && left.type() == right.type());
+
+        _disparity.create(size, CV_16UC1);
+        right_disp.create(size, CV_16UC1);
```

Comment:
For internal use, it might be better to be replaced `cv :: cuda :: ensureSizeIsEnough`.

Reply:
Done